### ISSUE TITLE
feat: DRM for IPTV

### ISSUE THREAD

**Issue by bgshankar1969** [rabilrbl/jiotv_go#334]
### Feature description

Please add DRM for IPTV links also. Most of the content on SONY channels is DRM protected. Normally Android TVs dont have a good browsers and depend on Kodi or some other IP TV solutions. 

So, please add DRM for IPTV links also. 

### Acknowledgements

- [X] I have searched the existing issues and this is a new and no duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I filled out all of the requested information in this issue properly.

**Comment by rabilrbl** [comment]
Hopefully someone from the community can work on this. I don't have the time to work on this
It also needs some modifications on existing DRM component.

**Comment by meghadeep-com** [comment]
First of all, thank you, contributors, for this amazing project! I had been looking for a solution to integrate Indian live TV channels on my media server for a while, and this fit the bill perfectly!

As for DRM support for IPTV, it works if you follow the rules of DRM media playback:

- DRM media is only played over secure connection (HTTPS, or HTTP over loopback interface). 
  - If you have a media server (e.g. Jellyfin, or Emby, or Plex), and if your JioTV Go installation is on the same host as your media server (same computer for bare metal binaries, or same podman pod if using containers), you can use localhost over http to set up IPTV channels on your media server, which will allow DRM media playback on all the clients that your media server serves over the network.
  - If your JioTV installation is on a different host than your media server, or if you do not have a media server, and connect to JioTV Go's IPTV playlists directly from IPTV clients, you'd need to get an SSL certificate using Let's Encrypt, and set up reverse proxy to handle secure connections.

I use both of these methods for a much easier playback. 

- I have Emby and Jellyfin in the same podman pod as I do the JioTV Go container in. So, I could set up IPTV on Emby Server and Jellyfin Server using `http://localhost:5001/playlist.m3u?q=high&l=English,Bengali`, and then use Emby client or Jellyfin client to play live TV on my Shield TV Pro over the network. DRM playback works with this method.
- I have also set up reverse proxy so that I can connect to JioTV Go directly on the internet. This lets me connect to JioTV Go directly at `https://jiotv.example.com` using any web browser, and since I am using valid SSL certs from Let's Encrypt, DRM playback works without any issues in web browsers on different machines.
  - During testing, before I set up the reverse proxy, I was connecting to JioTV Go from a different machine (e.g. http://192.168.7.25:5001/ on the browser), and DRM playback failed because I was not using secure connection. Once I moved to secure connections, DRM playback resumed.
  - Please note that JioTV Go does not have authentication, so you'd need to set up your own authentication in the reverse proxy itself (I use Authentik for that), or anyone on the internet will be able to watch your JioTV Go.

**Comment by rabilrbl** [comment]
@meghadeep-com Yeah, it’s mainly due to the architecture we use with DRM — it was put together quickly to support the web, but the downside is that it’s not very IPTV-friendly.
I also don’t have much experience with IPTV clients myself, since I don’t use them.
Would be great if someone with the right expertise and ability to test could take this up.

**Comment by itskagee** [comment]
Taking up this feature request.

Fix is almost ready. Just going to test it over the next week before I set up a PR.

**Comment by itskagee** [comment]
@rabilrbl I'm curious about something.

The main documentation and comments of the default configs say that we want DRM to be false by default. 
But in the code, under internal > handlers.go, you've specifically hard coded the DRM value to be `true`, overriding the user's config. Same for the actual values of the default configs.

The changes were made in PR #551 and #630.

Am I to assume that this is intended? If so I'll fix the documentation, configs and associated tests along with the other DRM changes I'm making.

**Comment by rabilrbl** [comment]
@itskagee I think I did that as people were complaining few channels didn't work.
The way it works currently is when upstream returns drm flag true, we automatically use it.

Yes, the documentation is out of date. Proceed with updating it.

But many new channels wouldn't work with HLS and some old channels have very low quality (360p) in HLS whereas DRM has better resolution.

**Comment by itskagee** [comment]
Yes. Almost 900 channels use DRM now as per my latest scrape.
And this is just for Hindi and English channels.

Alright, I'll make the associated changes.
I'll default the DRM to `true`, but instead of ignoring the user's config like we're doing now, I'll change it so that it is still taken into account, just in case someone wants to disable DRM (since not all IPTV apps support encrypted streams).

These users won't be able to play DRM streams anyway, so might as well give them the option to turn it off so that they can at least see some channels in low quality. For people using the web player, the default status will take care of it.

In the future I'm planning on implementing a background process that does this scraping once a day (or week, we'll discuss) to update any and all DRM channels as they come about, since the list may keep on changing.
How does this sound?
I'll put up a separate PR for that after this one.

For now, I'm just going to change the `SONY_LIST` to `DRM_LIST` and change the code so that it reads from a separate 'csv' file, which I'll populate with those 900 or so channels as a default. This should give us a good base to work on for later.

**Comment by rabilrbl** [comment]
Not sure I understand why we need a background process to update channels?
The channels are fetched real-time from Jio API.

Please keep SONY_LIST separate, it's required for some logic that used Sony servers for playback.

**Comment by itskagee** [comment]
So that real time fetching works for the web player, but IPTV apps are dumb. They require a static playlist file with predefined stream url and certain tags that identify DRM streams.

Those have to be precomputed since only DRM streams should have the DRM tags. If you put DRM tags on normal streams, they won't play. And since the tags are applied during playlist generation, it cannot be done in real time on the IPTV player.

One way to do so is checking for all DRM channels every time an IPTV app requests a playlist.But as you can imagine that'll take quite a long time since there have to be some reasonable delays between every DRM check to not hammer the API. No one wants to wait 20 minutes just to get their playlist.

Second way to do it is how I'm doing it right now, by pre-computing a file with all the channels beforehand. But this is static and we'd need to push an update every time there's a change in the DRM status of any channel.

The third way is for us to have a dynamic scraper that updates a gist here in the repo with any changes that occur, and our server reads the channel list from it. But this seems like an over complication to me, and comes with the drawback of automating it on Github, which won't be trivial.

The fourth option is the background process I proposed.

I could not think of a better approach, so I'd be interested in hearing your thoughts and whether you think there's a better to solve this.

As for the `SONY_LIST`, it is just being used here: [Line](https://github.com/JioTV-Go/jiotv_go/blob/4b63aa25b60d55fc8687b64ce7af1aad09f38724/internal/handlers/handlers.go#L1034)

It seems like the only thing it does is change the player URL to use 'MPD' streams, which is exactly what all the other DRM streams need as well. Hence my proposal of merging them.

**Comment by rabilrbl** [comment]
I forgot about DRM playlist,
You're right, I agree with 4th option having background process.
We can't have jiotv_go dependent on any other service besides Jio API.

As for SONY_LIST, go ahead and check if every channels work after change and it didn't break any flows.

**Comment by itskagee** [comment]
Already did!
Everything works and all existing tests are passing with that change.

**Comment by itskagee** [comment]
Just a note: Some Sony channels like Sony Ten 5 HD, Sony Ten 1 HD etc don't work in the current release as well (404 on both HLS and MPD streams), so this issue isn't related to the changes I'll be making to the list and the DRM.

**Comment by rabilrbl** [comment]
> Just a note: Some Sony channels like Sony Ten 5 HD, Sony Ten 1 HD etc don't work in the current release as well (404 on both HLS and MPD streams), so this issue isn't related to the changes I'll be making to the list and the DRM.

Are we missing something on the API, or did Jio update something there?
Let's investigate this on a separate issue.
